### DISCUSSION TITLE
How to add cors from an external file (if possible?)

### DISCUSSION THREAD

**Discussion by waliddoch** [fastapi/fastapi#4217] [Questions] [answered]
Hi, 
I am trying to create a `config/` folder where I want to add all the configs (obvious) like database, dev/prod, etc' - for making everything "neater" and organised. 
Is it possible to implement cors with an external file? 

referring to the [documentation](https://fastapi.tiangolo.com/tutorial/cors/), the cors need to be added like this (after defining "app" in `main.py`):

``` 
from fastapi.middleware.cors import CORSMiddleware
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

I created the file `configs/cors.py`
```
origins = [
    "*",
    # "something else?" # set origin 
]
allows = [
    f'allow_origins={str(origins)}',
    'allow_credentials=True',
    'allow_methods=["POST"]',
    'allow_headers=["A-GatewayID", "A-GatewayVersion"]'
]

cors = [CORSMiddleware, allows]
```

and in `main.py`:
```
from .configs import cors
app.add_middleware(cors)
```

but get an error which I'm almost sure I added the types badly, and changing this won't make it look more organised.  
Has anyone done that before? 
Maybe in a better neater way?

**Comment by simondale00** [comment, marked as answer]
Hey, @devwal. Can you share the error you're getting? That may help diagnose the issue.

Though at first glance it is likely related to the types of arguments you're passing in. The `add_middleware` function expects multiple arguments: the first argument is the middleware class, while the remaining arguments are keyword arguments representing options to configure the middleware. ([Reference](https://github.com/encode/starlette/blob/master/starlette/applications.py#L125)).

I can see two potential issues in the snippet above. First, `cors` is the only variable passed to the `add_middleware` function. It's a list rather than a class. The second issue is likely with the `allows` list. This is an array of strings, rather than a list of key/values. Changing the code as follows may solve your problem.

`configs/cors.py`
```
allows = {
    'allow_origins': origins,
    'allow_credentials': True,
    'allow_methods': ["POST"],
    'allow_headers': ["A-GatewayID", "A-GatewayVersion"]
}
```

`main.py`
```
from .configs import cors

app.add_middleware(CORSMiddleware, **cors.allows)
```

Best,
Simon
